src/index.ml

```diff
@@ -450,56 +496,82 @@ module Make (K : Key) (V : Value) (IO : IO) = struct
-    let generation = Int64.succ t.generation in
-    let log_array =
-      let compare_entry e e' = compare e.key_hash e'.key_hash in
-      let b = Array.make (Tbl.length log.mem) witness in
-      Tbl.fold
-        (fun key value i ->
-          b.(i) <- { key; value; key_hash = K.hash key };
-          i + 1)
-        log.mem 0
-      |> ignore;
-      Array.fast_sort compare_entry b;
-      b
-    in
-    let fan_size =
-      match t.index with
-      | None -> Tbl.length log.mem
-      | Some index ->
-          (Int64.to_int (IO.offset index.io) / entry_size) + Tbl.length log.mem
+    let log_async =
+      let io =
+        let log_async_path = log_async_path t.root in
+        IO.v ~fresh:true ~readonly:false ~generation:(Int64.succ t.generation)
+          ~fan_size:0L log_async_path
+      in
+      let mem = Tbl.create 0 in
+      { io; mem }
     in
-    let fan_out = Fan.v ~hash_size:K.hash_size ~entry_size fan_size in
-    let merge =
-      IO.v ~readonly:false ~fresh:true ~generation
-        ~fan_size:(Int64.of_int (Fan.exported_size fan_out))
-        merge_path
+    t.log_async <- Some log_async;
+
+    let go () =
+      let generation = Int64.succ t.generation in
+      let log_array =
+        let compare_entry e e' = compare e.key_hash e'.key_hash in
+        let b = Array.make (Tbl.length log.mem) witness in
+        Tbl.fold
+          (fun key value i ->
+            b.(i) <- { key; key_hash = K.hash key; value };
+            i + 1)
+          log.mem 0
+        |> ignore;
+        Array.fast_sort compare_entry b;
+        b
+      in
+      let fan_size =
+        match t.index with
+        | None -> Tbl.length log.mem
+        | Some index ->
+            (Int64.to_int (IO.offset index.io) / entry_size)
+            + Array.length log_array
+      in
+      let fan_out = Fan.v ~hash_size:K.hash_size ~entry_size fan_size in
+      let merge =
+        let merge_path = merge_path t.root in
+        IO.v ~fresh:true ~readonly:false ~generation
+          ~fan_size:(Int64.of_int (Fan.exported_size fan_out))
+          merge_path
+      in
+      let index =
+        match t.index with
+        | None ->
+            let io =
+              IO.v ~fresh:true ~readonly:false ~generation ~fan_size:0L
+                (index_path t.root)
+            in
+            append_remaining_log fan_out log_array 0 merge;
+            { io; fan_out }
+        | Some index ->
+            let index = { index with fan_out } in
+            merge_with log_array index merge;
+            index
+      in
+      Fan.finalize index.fan_out;
+      IO.set_fanout merge (Fan.export index.fan_out);
+      Mutex.lock t.rename_lock;
+      IO.rename ~src:merge ~dst:index.io;
+      t.index <- Some index;
+      let log_async = assert_and_get t.log_async in
+      IO.clear ~keep_generation:true log.io;
+      Tbl.iter (append_key_value log.io) log_async.mem;
+      IO.sync log.io;
+      IO.set_generation log.io generation;
```

Comment:
why setting the generation so late? would be better to do this just after the clear I think